repl/repl.go

```diff
@@ -57,8 +61,121 @@ func Start(in io.Reader, out io.Writer) {
 		prompt.OptionPrefix("⧐  "),
 		prompt.OptionLivePrefix(changeLivePrefix),
 		prompt.OptionTitle("abs-repl"),
+		prompt.OptionHistory(getHistory()),
 	)
 	p.Run()
+	// we get here on ^D from the prompt
+	saveHistory()
+}
+
+/*
+Support for abs history file in the interactive REPL:
+
+1) The current ABS_HISTORY_FILE is loaded into the prompt.Run() cycle
+   using prompt.OptionHistory(getHistory()). This also loads the local history as well.
+   Default ABS_HISTORY_FILE is "~/.abs_history".
+2) Append each non-null, unique next line passed from prompt to the executor() to the local history.
+   NB. the live prompt history shows duplicate next lines, but they are not saved to the local history.
+3) Save the local history whenever the prompt.Run() loop exits (^D) or the executor() exits (on quit).
+   Write the local history to the ABS_HISTORY_FILE up to ABS_MAX_HISTORY_LINES (default 1000 lines).
+
+Note that ABS_HISTORY_FILE and ABS_MAX_HISTORY_LINES variables may come from the OS environment.
+*/
+
+const (
+	ABS_HISTORY_FILE      = "~/.abs_history"
+	ABS_MAX_HISTORY_LINES = "1000"
+)
+
+var (
+	historyFilePath string   // fully expanded ABS_HISTORY_FILE path
+	historyMaxLines int      // ABS_MAX_HISTORY_LINES
+	history         []string // local history
+)
+
+// expand full path to ABS_HISTORY_FILE for current user and get ABS_MAX_HISTORY_LINES
+func expandHistoryFilePath() string {
+	// obtain any OS environment variablse
+	maxHistoryLines := os.Getenv("ABS_MAX_HISTORY_LINES")
+	if len(maxHistoryLines) == 0 {
+		maxHistoryLines = ABS_MAX_HISTORY_LINES
+	}
+	historyMaxLines, _ = strconv.Atoi(maxHistoryLines)
+	//
+	historyFile := os.Getenv("ABS_HISTORY_FILE")
+	if len(historyFile) == 0 {
+		historyFile = ABS_HISTORY_FILE
+	}
+	// identify the user's homeDir
+	user, ok := user.Current()
+	if ok != nil {
+		fmt.Println("Unable to resolve current userId for ~/.abs_history")
+		os.Exit(99)
+	}
+	// Default path to user's homeDir
+	path := user.HomeDir
+	// expand bash-style ~/ path prefix to homeDir (also works in windows)
+	if strings.HasPrefix(historyFile, "~/") {
+		path = strings.Replace(historyFile, "~/", path+"/", 1)
```

Comment:
BTW i've seen [this example](https://stackoverflow.com/questions/17609732/expand-tilde-to-home-directory) where it's recommended to use [filepath.Join](https://golang.org/pkg/path/filepath/#Join) -- check if that's something that would be applicable in here as well? We might then revisit the implementation of `cd()` as well if that's the case.

Actually, since we have similar code in `cd()` you might want to add a function in `util.go`, something like `ExpandPath(path string)` that you can reuse in both places.

Reply:
Ok, this works for me.

Reply:
Ok, this works for me also.